File: Algorithm/python/Main_12920_평범한_배낭_2.py
import sys

input = lambda: sys.stdin.readline().rstrip()

weight = []
sati = []
N, M = map(int, input().split())
dp = [0] * (M + 1)

for i in range(N):
    V, C, K = map(int, input().split())
    idx = 1

    while K > 0:
        tmp = min(idx, K)
        weight.append(V * tmp)
        sati.append(C * tmp)
        idx *= 2
        K -= tmp

for i in range(len(weight)):
    for j in range(M, 0, -1):
        if j >= weight[i]:
            dp[j] = max(dp[j], dp[j - weight[i]] + sati[i])
print(dp[M])

File: Algorithm/src/Main_12920_평범한_배낭_2.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main_12920_평범한_배낭_2 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    static int N, M;
    static int[] dp;
    static List<Integer> weight = new ArrayList<>();
    static List<Integer> sati = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        dp = new int[M + 1];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int V = Integer.parseInt(st.nextToken());
            int C = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());

            int idx = 1;
            while (K > 0) {
                int tmp = Math.min(idx, K);

                weight.add(V * tmp);
                sati.add(C * tmp);

                idx *= 2;
                K -= tmp;
            }
        }

        for (int i = 0; i < weight.size(); i++) {
            for (int j = M; j > 0; j--) {
                if (j >= weight.get(i)) {
                    dp[j] = Math.max(dp[j], dp[j - weight.get(i)] + sati.get(i));
                }
            }
        }

        bw.write(dp[M] + "");
        bw.close();
    }
}